src/Routers/UsersRouter.js

```diff
@@ -162,13 +182,20 @@ export class UsersRouter extends ClassesRouter {
             delete user.authData;
```

Comment:
all this block shoudl be put in the extracted method as well.

Reply:
So the issue here was if I move these two blocks to the _authenticateUserFromRequest function:

```
          // Sometimes the authData still has null on that keys
          // https://github.com/parse-community/parse-server/issues/935
          if (user.authData) {
            Object.keys(user.authData).forEach((provider) => {
              if (user.authData[provider] === null) {
                delete user.authData[provider];
              }
            });
            if (Object.keys(user.authData).length == 0) {
              delete user.authData;
            }
          }

          // Remove hidden properties.
          UsersRouter.removeHiddenProperties(user);
```

Three of PasswordPolicy tests for expiration will fail. It seems the order matters, here. The way I can see to solve it would be if I move this block into the _authenticateUserFromRequest as well so that the order is maintained, and so that this occurs before the authData check and removing hidden properties:

```
          // handle password expiry policy
          if (req.config.passwordPolicy && req.config.passwordPolicy.maxPasswordAge) {
            let changedAt = user._password_changed_at;

            if (!changedAt) {
              // password was created before expiry policy was enabled.
              // simply update _User object so that it will start enforcing from now
              changedAt = new Date();
              req.config.database.update('_User', { username: user.username },
                { _password_changed_at: Parse._encode(changedAt) });
            } else {
              // check whether the password has expired
              if (changedAt.__type == 'Date') {
                changedAt = new Date(changedAt.iso);
              }
              // Calculate the expiry time.
              const expiresAt = new Date(changedAt.getTime() + 86400000 * req.config.passwordPolicy.maxPasswordAge);
              if (expiresAt < new Date()) // fail of current time is past password expiry time
                throw new Parse.Error(Parse.Error.OBJECT_NOT_FOUND, 'Your password has expired. Please reset your password.');
            }
          }
```

Is it ok to have this entire chunk of code moved into _authenticateUserFromRequest so it is like so? 

```
          // handle password expiry policy
          if (req.config.passwordPolicy && req.config.passwordPolicy.maxPasswordAge) {
            let changedAt = user._password_changed_at;

            if (!changedAt) {
              // password was created before expiry policy was enabled.
              // simply update _User object so that it will start enforcing from now
              changedAt = new Date();
              req.config.database.update('_User', { username: user.username },
                { _password_changed_at: Parse._encode(changedAt) });
            } else {
              // check whether the password has expired
              if (changedAt.__type == 'Date') {
                changedAt = new Date(changedAt.iso);
              }
              // Calculate the expiry time.
              const expiresAt = new Date(changedAt.getTime() + 86400000 * req.config.passwordPolicy.maxPasswordAge);
              if (expiresAt < new Date()) // fail of current time is past password expiry time
                throw new Parse.Error(Parse.Error.OBJECT_NOT_FOUND, 'Your password has expired. Please reset your password.');
            }
          }

          // Sometimes the authData still has null on that keys
          // https://github.com/parse-community/parse-server/issues/935
          if (user.authData) {
            Object.keys(user.authData).forEach((provider) => {
              if (user.authData[provider] === null) {
                delete user.authData[provider];
              }
            });
            if (Object.keys(user.authData).length == 0) {
              delete user.authData;
            }
          }

          // Remove hidden properties.
          UsersRouter.removeHiddenProperties(user);

          delete user.password;
```

Reply:
I'll have a look because. Seems that we need to keep is split and clean.